### ISSUE TITLE
~~~Informal communication channel~~~

### ISSUE THREAD

**Issue by lexidor** [hhvm/xhp-js#6]
The last real commit to this project is from 2017. It requires hhvm `~3.6`. Is this project still useful enough to upgrade to modern day hack?

**Comment by fredemmott** [comment]
I think so in general - it provides a nice abstraction for client-side JS, and the need for that isn't really going to be addressed by changes in Hack.

That said, it's probably not worth updating this until XHP namespaces are supported

**Comment by connorgurney** [comment]
Out of curiosity, did anything ever come of this issue?

**Comment by lexidor** [comment]
> Out of curiosity, did anything ever come of this issue?

Xhp namespace support was introduced in September [see blog post](https://hhvm.com/blog/2020/09/02/XHP-namespaces-and-syntax.html). We haven't gotten around to updating this just yet. I have changed the issue title to reflect the status.

**Comment by connorgurney** [comment]
Cool beans, @lexidor—thanks! :grin:

**Comment by lexidor** [comment]
These changes will make working on this project in modern hhvm versions possible.
We still have some work ahead of us, but having the project pass the typechecker is a lot more inviting to work on.
I will try and get this on the agenda so we can figure out what still needs to be done.

Namespace, license, and React 16 are things I can think of right now, but there might be more.

**Comment by connorgurney** [comment]
Did you make any progress on this, @lexidor? Happy to help as required.

**Comment by lexidor** [comment]
Thanks a lot for checking in. I forgot about the one last thing to note on my PR #9. I did acknowledge it with a reply, so it wasn't on my radar anymore. I'll finish that PR ~this European evening~ tomorrow. Once #9 is merged, we could tackle the other issues (#7, #8, #10). The most the difficulty would be in checking if / how the React integration works #8. The license question can only be answered by someone @Facebook. Picking a namespace is a breeze, just prepare for bikeshedding. I'd pick `Facebook\XHP\JS`. We already had `Facebook\XHP\{ChildValidation, Core, HTML}` in facebook/xhp-lib.

**Comment by lexidor** [comment]
If you happen to use React, could you look at how modern React "compiles" jsx? We need to output this from `XHPReact`.

**Comment by lexidor** [comment]
> Did you make any progress on this, @lexidor? Happy to help as required.

I addressed the review comment. Once it is in master, work on the other issues can get started.

**Comment by connorgurney** [comment]
I'll be honest, I'm not sure whether to start, but I'm more than happy to give it my best shot!

**Comment by lexidor** [comment]
I'll have a look in the coming days about the code that React generates for jsx and how we can emit this from server side. It may very well be the case that React's core "codegen" has not changed much or at all. If so, this will be an easy task. A lot of best practices have changed on the web. Scripts are usually deferred by default, so we might need to still have this work in situations where React is not yet active during the first parse of the document.

**Comment by lexidor** [comment]
#9 got merged right before the weekend. Master can now run on hhvm 4.72 and up.

**Comment by connorgurney** [comment]
Superb news there! Moving forward, #7 needs to be addressed by @fredemmott or his colleagues at Facebook, and I think they'll want input on #10 too. I'm still happy to work on #8 with you, @lexidor.

**Comment by lexidor** [comment]
@connordoner that sounds great. I'll take a first look on a branch of mine tomorrow. Then I'll grant you access to that branch. If you prefer communicating via Github issues over email for example, let's do that on an issue that does not notify Fred. #8 seems like a good place for that. I'll close this issue, since it has been resolved by #9.

**Comment by connorgurney** [comment]
I'm more than happy to communicate however you prefer, @lexidor! My email is development@connorgurney.me.uk.

That said, it may well be worth keeping it in GitHub issues for the benefit of others using XHP?

**Comment by fredemmott** [comment]
Hey - we're also on Slack in #xhp - https://hhvm.com/slack :)

**Comment by connorgurney** [comment]
Thanks, Fred - I shall be over there shortly!
